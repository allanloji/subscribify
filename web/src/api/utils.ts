import { API_URL } from "@/utils/constants";
import { Api } from "./Api";

const api = new Api();
api.baseUrl = API_URL;

export { api };
