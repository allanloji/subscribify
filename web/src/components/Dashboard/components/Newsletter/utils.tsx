const backgrounds = [
  {
    primary: "#e1f0da",
    secondary: "#d4e7c5",
  },
  {
    primary: "#f6fedb",
    secondary: "#e6d3a3",
  },
  {
    primary: "#f9e3e3",
    secondary: "#f0c3c3",
  },
  {
    primary: "#e3e3f9",
    secondary: "#c3c3f0",
  },
  {
    primary: "#e3f9f9",
    secondary: "#c3f0f0",
  },
  {
    primary: "#f9e3f9",
    secondary: "#f0c3f0",
  },
  {
    primary: "#f9f3e3",
    secondary: "#f0d3c3",
  },
  {
    primary: "#f3f9e3",
    secondary: "#d3f0c3",
  },
  {
    primary: "#f3e3f9",
    secondary: "#d3c3f0",
  },
  {
    primary: "#e3f9e3",
    secondary: "#c3f0c3",
  },
  {
    primary: "#f9f3f3",
    secondary: "#f0d3d3",
  },
  {
    primary: "#98c1d9",
    secondary: "#6e9ebf",
  },
];

export const getBackground = (id: string) => {
  const sanitizedUuid = id.replace(/-/g, "");

  const uuidNumber = parseInt(sanitizedUuid.substring(0, 8), 16);

  const index = (uuidNumber * 10) % backgrounds.length;
  console.log(id, uuidNumber, index);

  return backgrounds[index];
};
