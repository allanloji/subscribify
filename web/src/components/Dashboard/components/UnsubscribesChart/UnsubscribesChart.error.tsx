import { Spacer } from "@/components/ui";
import * as S from "./UnsubscribesChart.styles";
import { Frown } from "lucide-react";

function UnsubscribesChartError() {
  return (
    <S.Container>
      <h2>Last 30 days unsubscribes</h2>
      <Spacer />
      <S.ErrorContainer>
        <span>Could not load the data. Please try again later</span>
        <Spacer size={2} />
        <Frown size={50} />
      </S.ErrorContainer>
    </S.Container>
  );
}

export default UnsubscribesChartError;
