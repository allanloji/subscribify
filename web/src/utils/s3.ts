import { v4 as uuid } from "uuid";
import {
  GetObjectCommand,
  PutObjectCommand,
  S3Client,
} from "@aws-sdk/client-s3";
import { getSignedUrl } from "@aws-sdk/s3-request-presigner";

const s3 = new S3Client({
  region: process.env.NEXT_PUBLIC_AWS_REGION || "",
  credentials: {
    accessKeyId: process.env.NEXT_PUBLIC_AWS_ACCESS_KEY_ID || "",
    secretAccessKey: process.env.NEXT_PUBLIC_AWS_SECRET_ACCESS_KEY || "",
  },
});

export async function uploadFile(file: File, name?: string) {
  const key = `${uuid()}.${file.name.split(".").pop()}`;
  const command = new PutObjectCommand({
    Bucket: process.env.NEXT_PUBLIC_AWS_BUCKET,
    Key: name || key,
    Body: file,
  });

  try {
    await s3.send(command);
    return key;
  } catch (err) {
    console.error(err);
  }
}

export async function getFile(key: string) {
  const command = new GetObjectCommand({
    Bucket: process.env.NEXT_PUBLIC_AWS_BUCKET,
    Key: key,
  });

  const url = await getSignedUrl(s3, command, { expiresIn: 3600 });

  return url;
}
