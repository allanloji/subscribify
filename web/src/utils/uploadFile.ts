import { v4 as uuid } from "uuid";
import { PutObjectCommand, S3Client } from "@aws-sdk/client-s3";

const s3 = new S3Client({
  region: process.env.NEXT_PUBLIC_AWS_REGION || "",
  credentials: {
    accessKeyId: process.env.NEXT_PUBLIC_AWS_ACCESS_KEY_ID || "",
    secretAccessKey: process.env.NEXT_PUBLIC_AWS_SECRET_ACCESS_KEY || "",
  },
});

export async function uploadFile(file: File) {
  const key = `${uuid()}.${file.name.split(".").pop()}`;
  const command = new PutObjectCommand({
    Bucket: process.env.NEXT_PUBLIC_AWS_BUCKET,
    Key: key,
    Body: file,
  });

  try {
    await s3.send(command);
    return key;
  } catch (err) {
    console.error(err);
  }
}
