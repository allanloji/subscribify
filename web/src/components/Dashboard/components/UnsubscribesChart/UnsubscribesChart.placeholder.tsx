import { Spacer } from "@/components/ui";
import Skeleton from "react-loading-skeleton";
import "react-loading-skeleton/dist/skeleton.css";
import * as S from "./UnsubscribesChart.styles";

function UnsubscribesChartPlaceholder() {
  return (
    <S.Container>
      <Skeleton height={24} width={300} />
      <Spacer />
      <Skeleton height={300} />
    </S.Container>
  );
}

export default UnsubscribesChartPlaceholder;
