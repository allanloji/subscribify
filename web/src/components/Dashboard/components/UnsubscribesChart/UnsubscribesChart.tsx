import { Spacer } from "@/components/ui";
import { API_URL } from "@/utils/constants";
import { useQuery } from "@tanstack/react-query";
import {
  AreaChart,
  Area,
  XAxis,
  YAxis,
  ResponsiveContainer,
  Tooltip,
} from "recharts";
import * as S from "./UnsubscribesChart.styles";
import UnsubscribesChartPlaceholder from "./UnsubscribesChart.placeholder";
import UnsubscribesChartError from "./UnsubscribesChart.error";

function UnsubscribesChart() {
  const {
    data: unsubscribesData,
    isLoading: isLoadingUnsubscribesData,
    isError: isErrorUnsubscribesData,
  } = useQuery({
    queryKey: ["unsubscribes"],
    queryFn: async () => {
      const response = await fetch(`${API_URL}/unsubscribes`, {
        headers: {
          "Content-Type": "application/json",
        },
      });
      return response.json();
    },
    select: (data) => {
      return data.map((d: any) => ({
        name: d.date,
        count: d.count,
      }));
    },
  });

  if (isLoadingUnsubscribesData) {
    return <UnsubscribesChartPlaceholder />;
  }

  if (isErrorUnsubscribesData) {
    return <UnsubscribesChartError />;
  }

  return (
    <S.Container>
      <h2>Last 30 days unsubscribes</h2>
      <Spacer />
      <ResponsiveContainer width="100%" height="90%">
        <AreaChart
          data={unsubscribesData}
          margin={{ top: 10, right: 30, left: 0, bottom: 0 }}
        >
          <defs>
            <linearGradient id="colorCount" x1="0" y1="0" x2="0" y2="1">
              <stop offset="5%" stopColor="#8884d8" stopOpacity={0.8} />
              <stop offset="95%" stopColor="#8884d8" stopOpacity={0} />
            </linearGradient>
          </defs>
          <XAxis dataKey="name" padding={{ right: 20 }} />
          <YAxis allowDecimals={false} />
          <Tooltip />
          <Area
            type="monotone"
            dataKey="count"
            stroke="#8884d8"
            fillOpacity={1}
            fill="url(#colorCount)"
          />
        </AreaChart>
      </ResponsiveContainer>
    </S.Container>
  );
}

export default UnsubscribesChart;
