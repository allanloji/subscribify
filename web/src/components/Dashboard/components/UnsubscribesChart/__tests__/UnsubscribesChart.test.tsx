import { beforeAll, afterEach, afterAll, expect, test } from "vitest";
import { server } from "@/utils/tests/server";
import { waitFor } from "@testing-library/react";
import { render } from "@/utils/tests/renderApp";
import { HttpResponse, http } from "msw";
import { API_URL } from "@/utils/constants";
import UnsubscribesChart from "..";

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

test("renders the unsubscribes placeholder", async () => {
  const { container } = render(<UnsubscribesChart />);

  expect(container).toMatchSnapshot();
});
