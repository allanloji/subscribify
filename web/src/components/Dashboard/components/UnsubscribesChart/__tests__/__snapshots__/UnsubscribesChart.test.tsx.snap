// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders the unsubscribes placeholder 1`] = `
<div>
  <section
    class="css-1ufao52"
  >
    <span
      aria-busy="true"
      aria-live="polite"
    >
      <span
        class="react-loading-skeleton"
        style="width: 300px; height: 24px;"
      >
        ‌
      </span>
      <br />
    </span>
    <div
      class="css-g1gnh4"
    />
    <span
      aria-busy="true"
      aria-live="polite"
    >
      <span
        class="react-loading-skeleton"
        style="height: 300px;"
      >
        ‌
      </span>
      <br />
    </span>
  </section>
</div>
`;
